react/index.d.ts

```diff
@@ -37,7 +37,7 @@ declare namespace React {
 
     type CElement<P, T extends Component<P, ComponentState>> = ComponentElement<P, T>;
     interface ComponentElement<P, T extends Component<P, ComponentState>> extends ReactElement<P> {
-        type: ComponentClass<P>;
+        type: ComponentClass<P> | ComponentClass<any>;
```

Comment:
Sounds like the right way to fix this would be to add an override to `cloneElement` that takes a single `element` argument with no `props` so that the compiler doesn't infer `undefined` for `Props`.